### ISSUE TITLE
t/re/pat_advanced_thr.t unstable?

### ISSUE THREAD

**Issue by toddr** [Perl/perl5#20420]
While releasing 5.37.5, I noticed several failures related to `t/re/pat_advanced_thr.t 

* http://perl.develop-help.com/raw/?id=286638
* https://perl5.test-smoke.org/report/5024754
* https://perl5.test-smoke.org/report/5024723
* https://www.nntp.perl.org/group/perl.daily-build.reports/2022/10/msg286614.html

Others are passing so maybe this isn't a big deal. I do notice several recent changes to the test which may have fixed the issue?

d44172c6022f8ca803930708edb35b6ba6f56833 and 91d110e02db3b8905c4c48d5fdd76a8439de31f3

**Comment by toddr** [comment]
This list of failures is not a complete list.

**Comment by jkeenan** [comment]
> 

Yes, I have been observing these failures for some time now.  You can look them up in the matrix here:
https://perl5.test-smoke.org/submatrix?test=../t/re/pat_advanced.t&pversion=5.37.5
https://perl5.test-smoke.org/submatrix?test=../t/re/pat_advanced_thr.t&pversion=5.37.5

The problem is that they are all **intermittent** failures, hence not easily reproducible.  They are overwhelmingly being reported from one smoke tester.  They have appeared regardless of C-compiler: gcc, g++ and clang are all implicated.  I began discussing this with @khwilliamson off-list about 10 days ago, and I believe the commits you cite are his attempt to grapple with the problem.  We'll have to follow the smoke-test logs to see if they are cleared up or reduced following today's release.

For what it's worth, the logs provided by the smoke-tester show results like this:


2. https://perl5.test-smoke.org/report/5024249
```
$ zgrep -B6 -A1 -E '(FAILED|Configuration:)' log35458d36d3c9fcf47064e4379e15276057b0bca5.log.gz |cat
[2022-10-09 17:52:19-0400] Running smoke tests without $ENV{PERLIO}
[2022-10-09 17:52:19-0400] Running test suite only with 'harness' with HARNESS_OPTIONS=j4.
[2022-10-09 17:52:19-0400] make distclean ...
[2022-10-09 17:52:20-0400] Smoking patch 35458d36d3c9fcf47064e4379e15276057b0bca5 v5.37.4-92-g35458d36d3
[2022-10-09 17:52:20-0400] Smoking branch blead
[2022-10-09 17:52:25-0400]
[2022-10-09 17:52:25-0400] Configuration: -Dusedevel -Dcc="ccache g++"
[2022-10-09 17:52:25-0400] ------------------------------------------------------------------------------
-- 
No saved state, selection will be empty
sh: warning: setlocale: LC_ALL: cannot change locale (This is not a legal locale name): No such file or directory
[2022-10-09 18:59:00-0400]
[2022-10-09 18:59:00-0400] All tests successful.
[2022-10-09 18:59:00-0400] Archived results...
[2022-10-09 18:59:00-0400]
[2022-10-09 18:59:00-0400] Configuration: -Dusedevel -Dcc="ccache g++" -DDEBUGGING
[2022-10-09 18:59:00-0400] ------------------------------------------------------------------------------
-- 
sh: warning: setlocale: LC_ALL: cannot change locale (This is not a legal locale name): No such file or directory
# Test process timed out - terminating
make: *** [makefile:856: test_harness] Error 1

error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/bin/../lib/Test/Smoke.pm line 195.
[2022-10-09 19:54:15-0400]
[2022-10-09 19:54:15-0400] ../t/re/pat_advanced.t......................................FAILED
[2022-10-09 19:54:15-0400]     No plan found in TAP output
[2022-10-09 19:54:15-0400] Archived results...
[2022-10-09 19:54:15-0400]
[2022-10-09 19:54:15-0400] Configuration: -Dusedevel -Dcc="ccache g++" -Duseithreads
[2022-10-09 19:54:15-0400] ------------------------------------------------------------------------------
-- 
# Test process timed out - terminating
# Test process timed out - terminating
make: *** [makefile:856: test_harness] Error 2

error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/bin/../lib/Test/Smoke.pm line 195.
[2022-10-09 20:55:01-0400]
[2022-10-09 20:55:01-0400] ../t/re/pat_advanced.t......................................FAILED
[2022-10-09 20:55:01-0400]     No plan found in TAP output
[2022-10-09 20:55:01-0400] ../t/re/pat_advanced_thr.t..................................FAILED
[2022-10-09 20:55:01-0400]     Non-zero exit status: 1
[2022-10-09 20:55:01-0400] ../t/re/pat_advanced_thr.t..................................FAILED
[2022-10-09 20:55:01-0400]     No plan found in TAP output
[2022-10-09 20:55:01-0400] Archived results...
[2022-10-09 20:55:01-0400]
[2022-10-09 20:55:01-0400] Configuration: -Dusedevel -Dcc="ccache g++" -Duseithreads -DDEBUGGING
[2022-10-09 20:55:01-0400] ------------------------------------------------------------------------------
-- 
sh: warning: setlocale: LC_ALL: cannot change locale (This is not a legal locale name): No such file or directory
# Test process timed out - terminating
make: *** [makefile:856: test_harness] Error 1

error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/bin/../lib/Test/Smoke.pm line 195.
[2022-10-09 21:54:52-0400]
[2022-10-09 21:54:52-0400] ../t/re/pat_advanced_thr.t..................................FAILED
[2022-10-09 21:54:52-0400]     Non-zero exit status: 1
[2022-10-09 21:54:52-0400] ../t/re/pat_advanced_thr.t..................................FAILED
[2022-10-09 21:54:52-0400]     No plan found in TAP output
```
If I read these correctly, what we are seeing is that the shell is emitting a warning about the C command `setlocale`' with category `LC_ALL`.  This is causing a timeout in one of those two test programs. The `"No plan found in TAP output"` is something which I think means that `done_testing()` was not reached from within a test program -- which would be the case of a timeout.

I suspect that this is the sort of test failure that occurs when a VM doing smoke-tests is too small or too slow to always get it right, but I have no way of proving that.  And since the test failures are intermittent, I can't really bisect the problem.

**Comment by khwilliamson** [comment]
This should be fixed by https://github.com/Perl/perl5/commit/dd8f4e6b1cdd0b2bce8bb0601054d575b69b4265

**Comment by jkeenan** [comment]
> This should be fixed by [dd8f4e6](https://github.com/Perl/perl5/commit/dd8f4e6b1cdd0b2bce8bb0601054d575b69b4265)

What I recommend is that we monitor the https://perl5.test-smoke.org/submatrix results for the two test programs that have been displaying these intermittent failures.  Those two URLs are:

https://perl5.test-smoke.org/submatrix?test=../t/re/pat_advanced.t

https://perl5.test-smoke.org/submatrix?test=../t/re/pat_advanced_thr.t

The submatrix reports results via the output of `git describe`.  The commit cited above is: `v5.37.5-55-gdd8f4e6b1c`.

As of the moment I write this, there are no failures in either of those programs past @khwilliamson's commit.  I recommend we monitor for 7 more days before closing this ticket.

**Comment by jkeenan** [comment]
> > This should be fixed by [dd8f4e6](https://github.com/Perl/perl5/commit/dd8f4e6b1cdd0b2bce8bb0601054d575b69b4265)
> 
> What I recommend is that we monitor the https://perl5.test-smoke.org/submatrix results for the two test programs that have been displaying these intermittent failures. Those two URLs are:
> 
> https://perl5.test-smoke.org/submatrix?test=../t/re/pat_advanced.t
> 
> https://perl5.test-smoke.org/submatrix?test=../t/re/pat_advanced_thr.t
> 
> The submatrix reports results via the output of `git describe`. The commit cited above is: `v5.37.5-55-gdd8f4e6b1c`.
> 
> As of the moment I write this, there are no failures in either of those programs past @khwilliamson's commit. I recommend we monitor for 7 more days before closing this ticket.

Monitored for 8 days; no new reports of failures at those two URLs.  Closing this ticket.
